### PR TITLE
Migration Assessment Tool

### PR DESCRIPTION
**Pull request by sanyamsinghal**

### CODE REVIEW

**Comment by sanyamsinghal** [comment]
https://jenkins.dev.yugabyte.com/job/users/job/yb-voyager-testing/job/yb-voyager-testing-pipeline/2188/

File: yb-voyager/src/migassessment/common.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+		}
+
+		result = append(result, &tmplRec)
+	}
+
+	return result, nil
+}
+
+func loadCSVDataFileGeneric(filePath string) ([]Record, error) {
+	file, err := os.Open(filePath)
```
Comment (amit-yb) [inline yb-voyager/src/migassessment/common.go:77]: If the file does not exists, then return `nil, nil`

File: yb-voyager/src/srcdb/data/yb-voyager-gather-metadata-and-stats.psql [hunk 2]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/data/yb-voyager-gather-metadata-and-stats.psql:1]: As suggested in one of the other comments:
- provide each query in a separate file;
- provide a bash script to run `psql SCRIPT_FILE` for each query;
- the bash script also calls `pg_dump` to dump schema in a single file. The command arguments should be same as used by the `export schema` command.

**Review by amit-yb — APPROVED** [review]
LGTM with a few minor comments. Good work! 👍

File: yb-voyager/cmd/assessMigrationCommand.go [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+	stdout, err := preparedPsqlCmd.CombinedOutput()
+	if err != nil {
+		log.Errorf("output of postgres metadata and stats gathering script\n%s", string(stdout))
+		return err
+	}
+	log.Infof("output of postgres metadata and stats gathering script\n%s", string(stdout))
+	return nil
+}
+
+func generateAssessmentReport() error {
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/cmd/assessMigrationCommand.go:245]: Thanks Amit! Nothing special, just the thought process and learnings from you :)

File: yb-voyager/src/migassessment/common.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	*ShardingReport `json:"sharding"`
+	*SizingReport   `json:"sizing"`
+}
+
+var assessmentParams = &AssessmentParams{}
+
+type AssessmentParams struct {
+	TargetYBVersion string `toml:"target_yb_version"`
+	ShardingParams  `toml:"sharding_params"`
+	SizingParams    `toml:"sizing_params"`
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/src/migassessment/common.go]: done

File: yb-voyager/cmd/assessMigrationCommand.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+		if startClean {
+			utils.CleanDir(filepath.Join(exportDir, "assessment", "data"))
+			utils.CleanDir(filepath.Join(exportDir, "assessment", "reports"))
+		} else {
+			utils.ErrExit("metadata or reports files already exist in the assessment directory at '%s'. ", assessmentDir)
+		}
+	}
+}
+
+func exportSchemaForAssessMigration() (err error) {
```
[the remark this answers is not in the document]
Comment (sanyamsinghal) [reply yb-voyager/cmd/assessMigrationCommand.go]: Done

File: yb-voyager/src/srcdb/data/yb-voyager-gather-metadata-and-stats.psql (hunk 2 above)
Comment (sanyamsinghal) [reply yb-voyager/src/srcdb/data/yb-voyager-gather-metadata-and-stats.psql]: change done

File: yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    echo "  schema_list            Comma-separated list of schemas for which statistics are to be collected."
+    echo "                         Example: 'public,sales,inventory'"
+    echo ""
+    echo "  assessment_data_dir    The directory path where the assessment data will be stored."
+    echo "                         This script will attempt to create the directory if it does not exist."
+    echo ""
+    echo "Example:"
+    echo "  yb-voyager-gather-assessment-data.sh 'postgresql://user:pass@localhost:5432/mydatabase' 'public,sales' '/path/to/assessment/data'"
+    echo ""
+    echo "Please ensure to replace the placeholders with actual values suited to your environment."
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh:37]:
```
[amit.jambure:~] % HELP_TEXT="
dquote> line1
dquote> line2
dquote> line3
dquote> "
[amit.jambure:~] % echo $HELP_TEXT

line1
line2
line3

[amit.jambure:~] %
```

File: yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    echo "Usage: $0 <pg_connection_string> <schema_list> <assessment_data_dir>"
+    exit 1
+fi
+
+pg_connection_string=$1
+schema_list=$2
+assessment_data_dir=$3
+
+SCHEMA_DIR=$(dirname $0)
+echo "SCHEMA_DIR = $SCHEMA_DIR and $(dirname $0)"
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh:57]: Why the part the ` and ...`?

File: yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+echo "Collect stats about size of tables"
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh:59]: `Collect table sizes`

File: yb-voyager/src/srcdb/postgres.go [hunk 9]
```diff
@@ -271,9 +271,15 @@ func (pg *PostgreSQL) GetConnectionUriWithoutPassword() string {
...
-	pg.checkSchemasExists()
-	pgdumpExtractSchema(pg.source, pg.GetConnectionUriWithoutPassword(), exportDir)
+func (pg *PostgreSQL) ExportSchema(exportDir string, schemaDir string) {
+	if utils.FileOrFolderExists(filepath.Join(schemaDir, "schema.sql")) {
+		// case for assess-migration cmd workflow
+		log.Infof("directly parsing the '%s/schema.sql' file", schemaDir)
+		parseSchemaFile(exportDir, schemaDir, pg.source.ExportObjectTypeList)
+	} else {
+		pg.checkSchemasExists()
+		pgdumpExtractSchema(pg.source, pg.GetConnectionUriWithoutPassword(), exportDir, schemaDir)
```
Comment (amit-yb) [inline yb-voyager/src/srcdb/postgres.go:281]: `pgdumpExtractSchema()` is doing two tasks: dumping the schema and then parsing it. Move the call to `parseSchemaFile()` out of the `pgdumpExtractSchema()`.

**Review by amit-yb — APPROVED** [review]
LGTM

File: yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh (hunk 7 above)
Comment (sanyamsinghal) [reply yb-voyager/src/srcdb/data/gather-assessment-data/postgresql/yb-voyager-gather-assessment-data.sh:57]: it was for debugging, will remove it.

File: yb-voyager/src/srcdb/postgres.go (hunk 9 above)
Comment (sanyamsinghal) [reply yb-voyager/src/srcdb/postgres.go:281]: done

**Pull request merged**